plutus-core/plutus-core/src/PlutusCore/Builtin/Result.hs

```diff
@@ -58,9 +58,19 @@ data BuiltinError
 -- logging, since there's no logging on the chain and builtins don't emit much anyway. Otherwise
 -- we'd have to use @text-builder@ or @text-builder-linear@ or something of this sort.
 data BuiltinResult a
-    = BuiltinFailure (DList Text) BuiltinError
-    | BuiltinSuccess a
+    = -- 'BuiltinSuccess' is the first constructor to make it a bit more likely for GHC to
+      -- branch-predict it (which is something that we want, because most builtins return this
+      -- constructor). It is however not guaranteed that GHC will predict it, because even though
+      -- it's likely going to be a recursive case (it certainly is in the CEK machine) and thus the
+      -- constructor has precedence over 'BuiltinFailure', it doesn't have precedence over
+      -- 'BuiltinSuccessWithLogs', since that case is equally likely to be recursive.
```

Comment:
I wonder what happens if instead of 
```
 | BuiltinSuccess a
 | BuiltinSuccessWithLogs (DList Text) a
```
we have
```
 | BuiltinSuccess (DList Text) a
```
with an empty `DList Text` in case when there are no logs?

Reply:
We'll need to convert it to a list and either match on the list before doing logging to see if we need to do logging in the first place, which is quite a bit of unnecessary allocation (first the `DList` together with a pointer to it, then the list) plus we need to perform the match which will probably confuse the branch predictor as well. Or we could do logging without checking whether the list is empty, but that's calling a function passed dynamically, so another bunch of allocation and pointer chasing.

I.e. `BuiltinResult` is specifically designed to avoid all of that in the most common case of success with no logs.

Do feel free to try it out though if you're curious.